File: problems/29-advanced-trie/001-design-add-search-words/python/word_dictionary.py
class WordDictionary:
    def __init__(self):
        self.children = {}
        self.word = False

    def addWord(self, word: str) -> None:
        node = self
        for ch in word:
            if ch not in node.children:
                node.children[ch] = WordDictionary()
            node = node.children[ch]
        node.word = True

    def search(self, word: str) -> bool:
        def dfs(node, index):
            if index == len(word):
                return node.word

            ch = word[index]
            if ch != '.':
                next_node = node.children.get(ch)
                return next_node is not None and dfs(next_node, index + 1)

            for next_node in node.children.values():
                if dfs(next_node, index + 1):
                    return True
            return False

        return dfs(self, 0)

File: problems/29-advanced-trie/001-design-add-search-words/java/WordDictionary.java
public class WordDictionary {
    private static class Node {
        Node[] next = new Node[26];
        boolean word;
    }

    private final Node root = new Node();

    public void addWord(String word) {
        Node current = root;
        for (char ch : word.toCharArray()) {
            int index = ch - 'a';
            if (current.next[index] == null) current.next[index] = new Node();
            current = current.next[index];
        }
        current.word = true;
    }

    public boolean search(String word) {
        return dfs(root, word, 0);
    }

    private boolean dfs(Node node, String word, int index) {
        if (index == word.length()) return node.word;

        char ch = word.charAt(index);
        if (ch != '.') {
            Node next = node.next[ch - 'a'];
            return next != null && dfs(next, word, index + 1);
        }

        for (Node next : node.next) {
            if (next != null && dfs(next, word, index + 1)) return true;
        }
        return false;
    }
}
